File: android/src/main/java/dev/overair/capacitor/Bundles.kt
package dev.overair.capacitor

import java.io.File
import java.io.IOException
import java.io.InterruptedIOException
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.util.zip.ZipInputStream

/**
 * Bytes onto the device and into a directory the local server can host.
 *
 * Everything here streams. The archive is hashed WHILE it downloads and
 * unzipped straight from disk, so a 60 MB bundle costs a buffer rather than
 * 60 MB of heap - which is the main thing native buys over doing this in the
 * webview.
 */
class Bundles(private val root: File) {

    class VerifyError(message: String) : IOException(message)
    class Cancelled : IOException("cancelled")

    /** Reported as bytes arrive, and again at each phase change. */
    fun interface Progress {
        fun report(state: String, bytes: Long, total: Long)
    }

    fun dirFor(id: String): File = File(root, id)

    /**
     * Download, verify, unpack.
     *
     * The digest is checked before a single file is written, so a bundle
     * that does not match never touches a directory the webview might later
     * serve and a failed download cannot leave a half-applied tree behind.
     *
     * `cancelled` is polled rather than interrupting the thread: a half-torn
     * stream leaves a partial file that looks like a resumable download and
     * is not one.
     */
    fun install(
        id: String,
        url: String,
        expected: String,
        cancelled: () -> Boolean = { false },
        progress: Progress = Progress { _, _, _ -> },
    ): Pair<File, Long> {
        root.mkdirs()
        val archive = File(root, "$id.zip.part")
        val target = dirFor(id)

        try {
            val digest = download(url, archive, cancelled, progress)

            progress.report("VERIFYING", archive.length(), archive.length())
            if (!digest.equals(expected, ignoreCase = true)) {
                throw VerifyError("digest mismatch: expected $expected, got $digest")
            }
            if (cancelled()) throw Cancelled()

            progress.report("UNPACKING", archive.length(), archive.length())
            // A previous half-written attempt is rubbish, not a head start:
            // the tree must be exactly what the archive says.
            target.deleteRecursively()
            target.mkdirs()
            val size = unzip(archive, target, cancelled)
            // The web root is the top of the archive. A zipped www folder puts
            // index.html one level down and opens to "Webpage not available".
            if (!File(target, "index.html").isFile) {
                throw VerifyError("no index.html at the top of the bundle; zip the folder's contents, not the folder")
            }
            return target to size
        } catch (error: Throwable) {
            target.deleteRecursively()
            throw error
        } finally {
            archive.delete()
        }
    }

    /** Streams to disk and hashes in the same pass. Returns the hex digest. */
    private fun download(
        url: String,
        into: File,
        cancelled: () -> Boolean,
        progress: Progress,
    ): String {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            connectTimeout = 30_000
            readTimeout = 60_000
            instanceFollowRedirects = true
        }
        try {
            if (connection.responseCode !in 200..299) {
                throw IOException("download failed: HTTP ${connection.responseCode}")
            }
            // -1 when the server sends no length, which a presigned URL for a
            // streamed object legitimately may not.
            val total = connection.contentLengthLong.coerceAtLeast(0L)
            val md = MessageDigest.getInstance("SHA-256")
            var read = 0L
            // Ten a second at most: a progress bar cannot show more, and each
            // one crosses the bridge.
            var lastReport = 0L

            connection.inputStream.use { input ->
                into.outputStream().use { output ->
                    val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
                    while (true) {
                        if (cancelled()) throw Cancelled()
                        val count = input.read(buffer)
                        if (count < 0) break
                        output.write(buffer, 0, count)
                        md.update(buffer, 0, count)
                        read += count
                        val now = System.currentTimeMillis()
                        if (now - lastReport >= 100) {
                            progress.report("DOWNLOADING", read, total)
                            lastReport = now
                        }
                    }
                }
            }
            progress.report("DOWNLOADING", read, total)
            return md.digest().joinToString("") { "%02x".format(it) }
        } catch (error: InterruptedIOException) {
            throw if (cancelled()) Cancelled() else error
        } finally {
            connection.disconnect()
        }
    }

    /** Returns the unpacked size in bytes. */
    private fun unzip(archive: File, target: File, cancelled: () -> Boolean): Long {
        var total = 0L
        ZipInputStream(archive.inputStream().buffered()).use { zip ->
            var entry = zip.nextEntry
            while (entry != null) {
                if (cancelled()) throw Cancelled()
                val destination = resolve(target, entry.name)
                if (entry.isDirectory) {
                    destination.mkdirs()
                } else {
                    destination.parentFile?.mkdirs()
                    destination.outputStream().use { out -> total += zip.copyTo(out) }
                }
                zip.closeEntry()
                entry = zip.nextEntry
            }
        }
        return total
    }

    /**
     * An archive is untrusted input even after its digest matches: the digest
     * proves it is the bundle the server named, not that the bundle is sane.
     * An entry called `../../databases/app.db` would otherwise be written
     * exactly there - the Zip Slip vulnerability.
     */
    private fun resolve(target: File, name: String): File {
        val destination = File(target, name)
        val canonicalTarget = target.canonicalPath + File.separator
        if (!destination.canonicalPath.startsWith(canonicalTarget)) {
            throw VerifyError("refusing archive entry outside the bundle: $name")
        }
        return destination
    }

    /** Delete every bundle except the ones named. Keeping the previous one is
     *  not sentiment: it is what a failed boot falls back to. */
    fun prune(keep: Set<String>) {
        root.listFiles()?.forEach { entry ->
            if (entry.isDirectory && entry.name !in keep) entry.deleteRecursively()
        }
    }

    fun removeAll() {
        root.deleteRecursively()
    }
}

File: ios/Sources/OverairPlugin/Bundles.swift
import CryptoKit
import Foundation
import ZIPFoundation

/// Bytes onto the device and into a directory the local server can host.
///
/// The archive streams to a temporary file and is hashed in chunks, so a
/// 60 MB bundle costs a buffer rather than 60 MB of heap. That, and being
/// able to run before the webview exists, is what native buys here.
///
/// Unzipping needs ZIPFoundation: iOS ships no public zip API. Foundation's
/// archive support writes zips (NSFileCoordinator `.forUploading`) but cannot
/// read one, and Apple Archive is a different container format.
public final class Bundles {

    public enum Failure: LocalizedError {
        case http(Int)
        case digest(expected: String, got: String)
        case unsafeEntry(String)

        public var errorDescription: String? {
            switch self {
            case .http(let code):
                return "download failed: HTTP \(code)"
            case .digest(let expected, let got):
                return "digest mismatch: expected \(expected), got \(got)"
            case .unsafeEntry(let name):
                return "refusing archive entry outside the bundle: \(name)"
            }
        }
    }

    private let root: URL
    private let session: URLSession

    public init(root: URL, session: URLSession = .shared) {
        self.root = root
        self.session = session
    }

    public func directory(for id: String) -> URL {
        root.appendingPathComponent(id, isDirectory: true)
    }

    /// Download, verify, unpack.
    ///
    /// The digest is checked before a single file is written, so a bundle
    /// that does not match never touches a directory the webview might later
    /// serve, and a failed download cannot leave a half-applied tree behind.
    public func install(id: String, url: URL, expected: String) async throws -> (URL, Int64) {
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)

        // download(from:) streams to a temporary file; the bytes never sit in
        // memory in one piece.
        let (temporary, response) = try await session.download(from: url)
        defer { try? FileManager.default.removeItem(at: temporary) }

        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw Failure.http(http.statusCode)
        }

        let digest = try sha256(of: temporary)
        guard digest.caseInsensitiveCompare(expected) == .orderedSame else {
            throw Failure.digest(expected: expected, got: digest)
        }

        let target = directory(for: id)
        // A previous half-written attempt is rubbish, not a head start: the
        // tree must be exactly what the archive says.
        try? FileManager.default.removeItem(at: target)
        do {
            try FileManager.default.createDirectory(at: target, withIntermediateDirectories: true)
            try FileManager.default.unzipItem(at: temporary, to: target)
            try rejectEscapes(in: target)
            return (target, try size(of: target))
        } catch {
            try? FileManager.default.removeItem(at: target)
            throw error
        }
    }

    /// Hashes in 1 MB chunks, so the archive never has to fit in memory.
    public func sha256(of file: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: file)
        defer { try? handle.close() }
        var hasher = SHA256()
        while let chunk = try handle.read(upToCount: 1_048_576), !chunk.isEmpty {
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    /// An archive is untrusted input even after its digest matches: the digest
    /// proves it is the bundle the server named, not that the bundle is sane.
    /// A symlink or a `..` entry could otherwise reach outside the directory.
    private func rejectEscapes(in directory: URL) throws {
        let base = directory.standardizedFileURL.path
        guard let walker = FileManager.default.enumerator(
            at: directory, includingPropertiesForKeys: [.isSymbolicLinkKey]) else { return }
        for case let url as URL in walker {
            let values = try url.resourceValues(forKeys: [.isSymbolicLinkKey])
            if values.isSymbolicLink == true {
                throw Failure.unsafeEntry(url.lastPathComponent)
            }
            if !url.standardizedFileURL.path.hasPrefix(base) {
                throw Failure.unsafeEntry(url.lastPathComponent)
            }
        }
    }

    private func size(of directory: URL) throws -> Int64 {
        var total: Int64 = 0
        guard let walker = FileManager.default.enumerator(
            at: directory, includingPropertiesForKeys: [.fileSizeKey]) else { return 0 }
        for case let url as URL in walker {
            let values = try? url.resourceValues(forKeys: [.fileSizeKey])
            total += Int64(values?.fileSize ?? 0)
        }
        return total
    }

    /// Delete every bundle except the ones named. Keeping the previous one is
    /// not sentiment: it is what a failed boot falls back to.
    public func prune(keep: Set<String>) {
        let entries = (try? FileManager.default.contentsOfDirectory(
            at: root, includingPropertiesForKeys: nil)) ?? []
        for entry in entries where !keep.contains(entry.lastPathComponent) {
            try? FileManager.default.removeItem(at: entry)
        }
    }

    public func removeAll() {
        try? FileManager.default.removeItem(at: root)
    }
}
